code/network/multimsgs.cpp

```diff
@@ -7676,74 +7768,111 @@ void process_NEW_countermeasure_fired_packet(ubyte *data, header *hinfo)
 
-	target_sig = (target) ? target->net_signature : 0;
+		if (fire_info->target == nullptr) {
+			return;
+		}
 
-	u_beam_info = (short)beam_info_index;
+		shooter_subsys_index = ship_get_index_from_subsys(fire_info->turret, OBJ_INDEX(fire_info->shooter));
 
-	if (bfi_flags & BFIF_IS_FIGHTER_BEAM) {
-		Assert( (bank_point >= 0) && (bank_point < UCHAR_MAX) );
-		subsys_index = (char)bank_point;
-	} else {
-		subsys_index = (char)ship_get_index_from_subsys(turret, OBJ_INDEX(shooter));
-	}
+		Assertion(shooter_subsys_index >= 0, "BEAM fired from unknown subsystem!");
+		Assertion(shooter_subsys_index < SHRT_MAX, "BEAM fired from a subsystem beyond max limits!");
 
-	Assert(subsys_index >= 0);
-	if (subsys_index < 0) {
-		return;
+		if (shooter_subsys_index < 0) {
+			return;
+		}
 	}
 
-	// swap the beam_info override info into little endian byte order
-	b_info.dir_a.xyz.x = INTEL_FLOAT(&override->dir_a.xyz.x);
-	b_info.dir_a.xyz.y = INTEL_FLOAT(&override->dir_a.xyz.y);
-	b_info.dir_a.xyz.z = INTEL_FLOAT(&override->dir_a.xyz.z);
-
-	b_info.dir_b.xyz.x = INTEL_FLOAT(&override->dir_b.xyz.x);
-	b_info.dir_b.xyz.y = INTEL_FLOAT(&override->dir_b.xyz.y);
-	b_info.dir_b.xyz.z = INTEL_FLOAT(&override->dir_b.xyz.z);
-
-	b_info.delta_ang = INTEL_FLOAT(&override->delta_ang);
-	b_info.shot_count = override->shot_count;
+	target_sig = (fire_info->target) ? fire_info->target->net_signature : 0;
 
-	for (int i = 0; i < b_info.shot_count; i++) {
-		b_info.shot_aim[i] = INTEL_FLOAT(&override->shot_aim[i]);
+	if (fire_info->target && fire_info->target_subsys) {
+		target_subsys_index = ship_get_index_from_subsys(fire_info->target_subsys, OBJ_INDEX(fire_info->target));
 	}
 
+	u_beam_info = static_cast<short>(fire_info->beam_info_index);
+
 	// build the header
 	BUILD_HEADER(BEAM_FIRED);
-	ADD_USHORT(shooter->net_signature);
-	ADD_DATA(subsys_index);
+
+	ADD_USHORT(fire_info->shooter->net_signature);
 	ADD_USHORT(target_sig);
 	ADD_SHORT(u_beam_info);
-	ADD_DATA(b_info);  // FIXME: This is still wrong, we shouldn't be sending an entire struct over the wire - taylor
-//	ADD_DATA(bfi_flags);	// this breaks the protocol but is here in case we decided to do that in the future - taylor
-//	ADD_DATA(target_pos);	// ditto - Goober5000
+	ADD_FLOAT(fire_info->accuracy);
+	ADD_INT(fire_info->bfi_flags);
+	ADD_SHORT(static_cast<short>(shooter_subsys_index));
+	ADD_SHORT(static_cast<short>(target_subsys_index));
+
+	if (fire_info->bfi_flags & BFIF_IS_FIGHTER_BEAM) {
+		ADD_SHORT(static_cast<short>(fire_info->bank));
+		ADD_SHORT(static_cast<short>(fire_info->point));
+	}
+
+	// override info
+	{
+		// dir_a
+		ADD_FLOAT(override->dir_a.xyz.x);
+		ADD_FLOAT(override->dir_a.xyz.y);
+		ADD_FLOAT(override->dir_a.xyz.z);
+		// dir_b
+		ADD_FLOAT(override->dir_b.xyz.x);
+		ADD_FLOAT(override->dir_b.xyz.y);
+		ADD_FLOAT(override->dir_b.xyz.z);
+
+		ADD_FLOAT(override->delta_ang);
+
+		ADD_DATA(override->shot_count);
```

Comment:
As mentioned in the PR, (after the PR) shot_count won't need to be sent either.

Reply:
You'll find that counting on tabled information on packet receipt is not as reliable as it would seem.  While the risk here is much lower than in the object update packet, relying on the number of shield quadrants was causing multi crashes for years.  I couldn't rely on the subsystem information either.  

This is probably safe, just be cautious about your assumptions in multi, because they can cause mysterious crashes.

Reply:
Also, with IPv6 packet sizes, the shot_count would be only 1/1000th of the packet.

Reply:
I see, well I'll defer to your guys' judgment on this, but my chief concern was reducing the number of things that need to be sent. I am really quite worried if table data can't be relied upon, how on earth can that be the case?

Reply:
FSO is a minefield and things are just not all completely initialized exactly how you expect -- or at least that's what I've gathered.  Even though a ship is created, its subsystems may not be completely created or maybe it's just not put into the linked subsystem list yet.  

Or maybe there's a difference between the two computers because someone made some manual changes.  Yeah, their stats won't be saved on a PXO game, but no one is going to care, and then the beams you see on the client and server will differ.  

I honestly don't know enough about beams to say anything about this situation, I'm just saying I made this mistake early on.

Reply:
I should explain.  The server runs the simulation, and so the more information we can get from the server about how everything runs, the better.

Reply:
Fair enough, I'd like to see what taylor says though.

Reply:
Yeah, the shot count still needs to be added to the packet. With your changes it will only be used for making sure we get exactly the amount of data out of the packet as was put in though.

It's just not safe to depend on the client and server having the exact same data. For instance, if `MAX_BEAM_SHOTS` was changed then it could break multi if we didn't have the safety of the shot count in the packet.